### ISSUE TITLE
Feature request: Server initiated "window/showTextDocument" notification

### ISSUE THREAD

**Issue by rasika** [microsoft/language-server-protocol#612]
I would like to request a feature which is a **server-initiated** notification `window/showTextDocument` to open and/or show text document with a given [Location](https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md#location).

**Propose Change**
- a new method: 'window/showTextDocument'
- params: `Location` defined as follows:

```java
interface Location {
	uri: DocumentUri;
	range: Range;
}
```

**Why?**
This is beneficial when text files are generated by the Server and wants clients to open / show the text document(optionally to focus specific position) in cases such as **test files generation**, stubs generation.

I am already aware of the `CreateFile ` and `CreateFileOptions` still it won't allow to focus on specific position for a Server initiated notification(please correct me if I am wrong here).

WDYT?

**Comment by dbaeumer** [comment]
Makes sense to me. @rasika would you be will on working on a PR for this. I can assist and there is a guideline on how to do this: https://github.com/Microsoft/language-server-protocol/blob/master/contributing.md

I would make it a request to be able to return an indication if the opening was successful or failed.

**Comment by rasika** [comment]
@dbaeumer Thanks for the info. Sure, I'll proceed with the PR.

**Comment by dbaeumer** [comment]
@rasika just to clarify. We need a PR here and in https://github.com/Microsoft/vscode-languageserver-node since at least one implementation is necessary as well.

**Comment by rcjsuen** [comment]
Would it make sense to send an array so that you can open multiple files or just send multiple requests?

**Comment by dbaeumer** [comment]
I am fine sending an array.

**Comment by rasika** [comment]
@rcjsuen +1 for the array. Will implement accordingly.

**Comment by matklad** [comment]
I am personally not a huge fan of serve-initiated requests to change the state of the client, because they break the causality: you can't reliably infer what was the cause of the request. 

I'd rather we add an ability to focus specific document when applying edits. @rasika would it work for your if `WorkspaceEdit` create a `focus` array which described documents and locations editor should reveal after applying the edit?

@dbaeumer I would also appreciate an ability to focus a specific position after a document edit, and a specific document after workspace edit. In my rust-server, I use a [bespoke implementation](https://github.com/rust-analyzer/rust-analyzer/blob/82e3ab02afee7cc0db178e1b10ee65146e4b7a14/editors/code/src/commands/apply_source_change.ts#L48-L69) to have this feature. I need it in two cases:

* I have "create module" quick fix, which creates a new file, which better be focused
* SIngle-file code actions would like to move cursor to a specific position in the newly generated code.

**Comment by dbaeumer** [comment]
@matklad I would suggest to track this in another issue.

@rasika Could we make this a request instead of a notification. The open for example could fail for some reason.

**Comment by matklad** [comment]
@rasika could you clarify if you need a completely separate request to show arbitrary text document, or if you have a more specific requirement of "open a text document after creating it using `CreateFile` resource operation"?

**Comment by rasika** [comment]
Replied to https://github.com/Microsoft/language-server-protocol/issues/663; Adding a comment here for the future reference.

**Comment by darkbasic** [comment]
@rasika any news on the PR?
I'm adding resolvers finding capabilities to the GraphQL for VSCode extension (https://youtu.be/qBh6-Pd5Otw) and I definitely need this.
As you can see, in the video I'm still using the Go To Definition capabilities so everything's fine, but I've recently moved to `CodeLens` and I've found no way to initiate a showTextDocument from the server.
![0](https://user-images.githubusercontent.com/1047358/56228205-875e7500-6077-11e9-9711-f3332f2aeeb1.png)
Basically I have a `Location` object for each resolver and when the user clicks on `Go to Resolver` I want VSCode to open it and focus on it.

**Comment by darkbasic** [comment]
Since @rasika didn't answer I decided to give it a try: https://github.com/Microsoft/vscode-languageserver-node/pull/490

You're gonna use it this way:
```ts
await connection
  .sendRequest(ShowTextDocumentRequest.type.method, {
    fileName,
    options: {
      selection,
  })
```

I just implemented it for the node server without making any protocol proposal, because I'd like to discuss it little bit to see if there is anything that you would like to change before.

**Comment by rasika** [comment]
Apology for the late reply. Please go ahead. Won't be able to actively contribute due to my day job.

**Comment by darkbasic** [comment]
I just signed the CLA, let me know if there is anything else I can do to get such feature merged. We rely on it to release the next version of the graphql-for-vscode extension.

**Comment by KamasamaK** [comment]
The actual client capability does not appear to have been added.

**Comment by KamasamaK** [comment]
It happened again, so I've just opened a PR #1138 instead of posting another comment in another closed issue.

**Comment by dbaeumer** [comment]
@KamasamaK thanks. Merged the PR.
